src/System.Private.Reflection.Core/src/System/Reflection/Runtime/General/MetadataReaderExtensions.Ecma.cs

```diff
@@ -78,5 +86,257 @@ public static RuntimeAssemblyName ToRuntimeAssemblyName(this AssemblyReferenceHa
+                }
+
+                ctorType = memberRef.Parent;
+            }
+            else if (attributeCtor.Kind == HandleKind.MethodDefinition)
+            {
+                var methodDef = metadataReader.GetMethodDefinition((MethodDefinitionHandle)attributeCtor);
+
+                if (!metadataReader.StringComparer.Equals(methodDef.Name, ConstructorInfo.ConstructorName))
+                {
+                    // Not a constructor.
+                    throw new BadImageFormatException();
+                }
+
+                ctorType = methodDef.GetDeclaringType();
+                Debug.Assert(!ctorType.IsNil);
+            }
+            else
+            {
+                // invalid metadata
+                throw new BadImageFormatException();
+            }
+        }
+
+
+        /// <summary>
+        /// Given a token for a type, return the type's name and namespace.  Only works for top level types. 
+        /// namespaceHandle will be NamespaceDefinitionHandle for defs and StringHandle for refs. 
+        /// </summary>
+        /// <returns>True if the function successfully returns the name and namespace.</returns>
+        public static bool GetAttributeNamespaceAndName(MetadataReader metadataReader, EntityHandle typeDefOrRef, out StringHandle namespaceHandle, out StringHandle nameHandle)
+        {
+            nameHandle = default(StringHandle);
+            namespaceHandle = default(StringHandle);
+
+            if (typeDefOrRef.Kind == HandleKind.TypeReference)
+            {
+                TypeReference typeRefRow = metadataReader.GetTypeReference((TypeReferenceHandle)typeDefOrRef);
+                HandleKind handleType = typeRefRow.ResolutionScope.Kind;
+
+                if (handleType == HandleKind.TypeReference || handleType == HandleKind.TypeDefinition)
+                {
+                    // TODO - Support nested types.  
+                    return false;
+                }
+
+                nameHandle = typeRefRow.Name;
+                namespaceHandle = typeRefRow.Namespace;
+            }
+            else if (typeDefOrRef.Kind == HandleKind.TypeDefinition)
+            {
+                var def = metadataReader.GetTypeDefinition((TypeDefinitionHandle)typeDefOrRef);
+
+                if (IsNested(def.Attributes))
+                {
+                    // TODO - Support nested types. 
+                    return false;
+                }
+
+                nameHandle = def.Name;
+                namespaceHandle = def.Namespace;
+            }
+            else
+            {
+                // unsupported metadata
+                throw new BadImageFormatException();
+            }
+
+            return true;
+        }
+
+
+        public static bool IsNested(TypeAttributes flags)
+        {
+            return (flags & ((TypeAttributes)0x00000006)) != 0;
+        }
+
+        internal enum CorElementType : byte
+        {
+            Invalid,
+            ELEMENT_TYPE_VOID,
```

Comment:
Use verbatim copy of https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Reflection.Metadata/src/System/Reflection/Metadata/Internal/CorElementType.cs#L7 ?